{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b4ad9ae3",
   "metadata": {},
   "source": [
    "# Lasso Regression for Distance Analysis\n",
    "### Laurence Nickel (i6257119)\n",
    "\n",
    "Libraries used: \n",
    "* pandas (version: '1.2.4')\n",
    "* re (version: '2.2.1')\n",
    "* sys (version: '3.8.8')\n",
    "* os (version: '3.8.8')\n",
    "* time (version: '3.8.8')\n",
    "* plotly.express (version: '5.13.1')\n",
    "* seaborn (version: '0.11.1')\n",
    "* matplotlib.pyplot (version: '3.3.4')\n",
    "* sklearn.linear_model (version: '0.24.1')\n",
    "* sklearn.metrics (version: '0.24.1')\n",
    "* joblib (version: '1.0.1')\n",
    "\n",
    "References:\n",
    "* [1] Ranstam, J., \\& Cook, J. (2018). LASSO regression. *British Journal of Surgery, 105*(10), 1348. doi: https://doi.org/10.1002/bjs.10895.\n",
    "* [2] Hong, J., \\& Rhee, J. (2022). Genomic Effect of DNA Methylation on Gene Expression in Colorectal Cancer. *Biology (Basel), 11*(10): 1388. doi: 10.3390/biology11101388.\n",
    "* [3] Miles, J. (2005). \"R-squared, Adjusted R-squared,\" in *Encyclopedia of Statistics in Behavioral Science - Volume 4*, eds B. S. Everitt \\& D. C. Howell (Hoboken, NJ, USA: John Wiley \\& Sons), 1655-1657. doi: https://doi.org/10.1002/0470013192.bsa526."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afe594d5",
   "metadata": {},
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8976bd51",
   "metadata": {},
   "source": [
    "Within this notebook, the machine learning algorithm lasso regression is performed to predict the expression levels for the genes considering the methylation values for the CpG sites where different distances are experimented with the goal of finding which of the distances is the best to use, which denotes which CpG to use based on their location with respect to that gene, to predict the gene expression values. The dataset combination that is used to achieve this is the one with the M-transformed methylation data and the log2-transformed gene expression data. This was determined to be the best performing dataset combination within the notebook 'Linear Regression for Testing the Datasets.ipynb'. \n",
    "\n",
    "Lasso (Least Absolute Shrinkage and Selection Operator) regression is a linear regression method that uses a regularization term to perform variable selection and shrinkage of coefficient estimates [1]. Just like linear regression, lasso regression assumes a linear relationship between the independent variables and the dependent variable, but by including polynomials, it can also effectively model non-linear relationships. A line of best fit is found to relate a dependent variable to one or more independent variables, a linear equation that minimizes the sum of squared residuals. This is presented in Equation 1 where $Y$ represents the predicted gene expression value, $\\beta_0$ is the intercept (bias term), $\\beta_1, \\beta_2, \\ldots, \\beta_n$ are the coefficients corresponding to each CpG site, and $\\epsilon$ denotes the error term.\n",
    "\n",
    "<br></br>\n",
    "\\begin{equation}\n",
    "Y = \\beta_0 + \\beta_1 \\cdot CpG_1 + \\beta_2 \\cdot CpG_2 + \\ldots + \\beta_n \\cdot CpG_n + \\epsilon\\tag{1}\n",
    "\\end{equation}\n",
    "<br></br>\n",
    "\n",
    "The method Ordinary Least Squares (OLS) can be employed to estimate the coefficients $\\beta_1, \\beta_2, \\ldots, \\beta_n$. The OLS estimator aims to reduce the squared sum of differences between the observed and predicted values for gene expression. To this OLS estimator utilized by the linear regression, lasso regularization is added. This is presented in Equation 2 where $\\hat{\\beta}$ represents the estimated coefficient values that minimize the sum of squared errors and $N$ the number of samples.\n",
    "\n",
    "<br></br>\n",
    "\\begin{equation}\n",
    "\\hat{\\beta} = \\arg\\min_{\\beta} \\sum_{i=1}^N (Y_i - \\beta_0 - \\beta_1 \\cdot CpG_{1i} - \\beta_2 \\cdot CpG_{2i} - \\ldots - \\beta_n \\cdot CpG_{ni})^2 + \\lambda \\sum_{j=1}^{n} |\\beta_j|\\tag{2}\n",
    "\\end{equation}\n",
    "<br></br>\n",
    "\n",
    "The obtain the coefficients $\\beta_1, \\beta_2, \\ldots, \\beta_n$, the optimization problem in Equation 2 can be solved. Here, the $n$ represents the number of CpG sites, and the regularization parameter $\\lambda$ controls the regularization strength. The term $\\lambda \\sum_{j=1}^{n} |\\beta_j|$ introduces a penalty on the absolute values of the coefficients. By adjusting the value of $\\lambda$, lasso regression can perform feature selection by forcing some coefficients to be precisely zero. This indicates that the CpG sites corresponding to those coefficients have no significant impact on the prediction of gene expression values. \n",
    "\n",
    "Regarding the suitability of lasso regression for this thesis, it can be applied to predict gene expression values from methylation data as it has already been successfully performed before, showing that lasso regression is suitable for working with gene expression and methylation data [2]. Please mind that this does not mean that lasso regression is necessarily the best performing (regression) method for predicting gene expression values from methylation data, but applying the algorithm might provide us with reasonable results for our purpose of finding the distance with which the dataset combination performs the best.\n",
    "\n",
    "There are multiple distances (40 in total) that we will experiment with throughout this notebook. These are presented in the overview below:\n",
    "* 5,000\n",
    "* 10,000\n",
    "* 15,000\n",
    "* 25,000\n",
    "* 50,000\n",
    "* 75,000\n",
    "* 100,000\n",
    "* 150,000\n",
    "* 250,000\n",
    "* 350,000\n",
    "* 500,000\n",
    "* 750,000\n",
    "* 1,000,000\n",
    "* 1,500,000\n",
    "* 2,000,000\n",
    "* 2,500,000\n",
    "* 4,000,000\n",
    "* 5,000,000\n",
    "* 6,000,000\n",
    "* 7,500,000\n",
    "* 10,000,000\n",
    "* 12,500,000\n",
    "* 15,000,000\n",
    "* 17,500,000\n",
    "* 20,000,000\n",
    "* 25,000,000\n",
    "* 30,000,000\n",
    "* 40,000,000\n",
    "* 50,000,000\n",
    "* 65,000,000\n",
    "* 80,000,000\n",
    "* 100,000,000\n",
    "* 120,000,000\n",
    "* 150,000,000\n",
    "* 200,000,000\n",
    "* 250,000,000\n",
    "* 350,000,000\n",
    "* 500,000,000\n",
    "* 750,000,000\n",
    "* 1,000,000,000\n",
    "\n",
    "For each of these distances lasso regression models are built, one for each gene, and these are evaluated. The distance of the best performing experiment will represent which distance we should use to determine which CpG sites to consider (based on their position relative to a particular gene) for predicting the gene expression value of a particular gene.\n",
    "\n",
    "Unlike we did in the notebook 'Linear Regression for Testing the Datasets.ipynb', we will not only build models for the genes which are located on chromosome 1, which was done to reduce the computational burden, but instead we will use all of the genes present within the log2-transformed gene expression data since this notebook (among the other machine learning algorithms notebooks within this directory) represents the main experiments of the Distance Analysis part.\n",
    "\n",
    "To retrieve the prediction accuracy of the gene expression values, the R-squared (R<sup>2</sup>) metric is computed for each of the predictions which indicates the proportion of the variance in the dependent variable that can be explained by the model [3]. Higher R<sup>2</sup> values indicate a more significant proportion of the variance in the dependent variable that can be explained by the model, with 1 being the largest possible value. This R<sup>2</sup> value is retrieved by applying 4-fold cross-validation using the training and test splits defined in the notebook 'Training and Test Set Division.ipynb' present in the 'Machine Learning Algorithms - Preprocessing' folder, which also includes the motivation behind choosing the k in k-fold cross-validation to be set equal to four, and averaging the R<sup>2</sup> value for each of the four folds."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6eefe66",
   "metadata": {},
   "source": [
    "### Importing libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe822e4c",
   "metadata": {},
   "source": [
    "Before we can start to define all the functions, we should first import some libraries that will be used throughout this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf57a5f9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:01:00.419577Z",
     "start_time": "2023-06-16T21:00:43.464899Z"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting the importing of the libraries...\n",
      "Requirement already satisfied: plotly in c:\\users\\laure\\anaconda3\\lib\\site-packages (5.13.1)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in c:\\users\\laure\\anaconda3\\lib\\site-packages (from plotly) (8.2.2)\n",
      "Requirement already satisfied: joblib in c:\\users\\laure\\anaconda3\\lib\\site-packages (1.0.1)Finishing the installing of the libraries.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Starting the importing of the libraries...\")\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import re\n",
    "import sys\n",
    "import os\n",
    "import time\n",
    "\n",
    "# Here we first need to install the plotly library.\n",
    "!pip install plotly\n",
    "import plotly\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import sklearn\n",
    "from sklearn.linear_model import Lasso\n",
    "from sklearn.metrics import r2_score\n",
    "\n",
    "!pip install joblib\n",
    "import joblib\n",
    "from joblib import Parallel, delayed\n",
    "\n",
    "\n",
    "print(\"Finishing the installing of the libraries.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10fc1b6a",
   "metadata": {},
   "source": [
    "Now that all the libraries have been imported, we can verify that these libraries have been loaded into this notebook by calling the version property of the library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ff6c2bfb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:01:00.451483Z",
     "start_time": "2023-06-16T21:01:00.424718Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The library 'pd' (pandas) has been loaded into the notebook with its version being:\n",
      "1.2.4\n",
      "\n",
      "The library 're' has been loaded into the notebook with its version being:\n",
      "2.2.1\n",
      "\n",
      "The library 'sys' has been loaded into the notebook with its version being:\n",
      "3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]\n",
      "\n",
      "The library 'plotly' has been loaded into the notebook with its version being:\n",
      "5.13.1\n",
      "\n",
      "The library 'sns' (seaborn) has been loaded into the notebook with its version being:\n",
      "0.11.1\n",
      "\n",
      "The library 'matplotlib' has been loaded into the notebook with its version being:\n",
      "3.3.4\n",
      "\n",
      "The library 'sklearn' has been loaded into the notebook with its version being:\n",
      "0.24.1\n",
      "\n",
      "The library 'joblib' has been loaded into the notebook with its version being:\n",
      "1.0.1\n"
     ]
    }
   ],
   "source": [
    "# Retrieving the version of the libraries to verify they have been correctly loaded into this notebook.\n",
    "print(\"The library 'pd' (pandas) has been loaded into the notebook with its version being:\")\n",
    "print(pd.__version__)\n",
    "\n",
    "print(\"\\nThe library 're' has been loaded into the notebook with its version being:\")\n",
    "print(re.__version__)\n",
    "\n",
    "print(\"\\nThe library 'sys' has been loaded into the notebook with its version being:\")\n",
    "print(sys.version)\n",
    "\n",
    "print(\"\\nThe library 'plotly' has been loaded into the notebook with its version being:\")\n",
    "print(plotly.__version__)\n",
    "\n",
    "print(\"\\nThe library 'sns' (seaborn) has been loaded into the notebook with its version being:\")\n",
    "print(sns.__version__)\n",
    "\n",
    "print(\"\\nThe library 'matplotlib' has been loaded into the notebook with its version being:\")\n",
    "print(matplotlib.__version__)\n",
    "\n",
    "print(\"\\nThe library 'sklearn' has been loaded into the notebook with its version being:\")\n",
    "print(sklearn.__version__)\n",
    "\n",
    "print(\"\\nThe library 'joblib' has been loaded into the notebook with its version being:\")\n",
    "print(joblib.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d06ce1d1",
   "metadata": {},
   "source": [
    "### Defining the data directories"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ffbfaa8",
   "metadata": {},
   "source": [
    "In addition, we also need to define our data directories from which the gene expression and methylation data files and the training and test splits data files will be loaded. Please mind that these need to be changed to the desired directories to be able to work with the data directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ffd2f6f0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:01:00.467060Z",
     "start_time": "2023-06-16T21:01:00.457555Z"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_directory_final_datasets = \"C:/Users/laure/OneDrive/Documenten/Bachelor Thesis Data/final_datasets\"\n",
    "data_directory_final_datasets_distance = \"C:/Users/laure/OneDrive/Documenten/Bachelor Thesis Data/final_datasets/Distance Analysis\"\n",
    "data_directory_training_and_test_splits = \"C:/Users/laure/OneDrive/Documenten/Bachelor Thesis Data/training_and_test_splits\"\n",
    "data_directory_results_distance = \"C:/Users/laure/OneDrive/Documenten/Bachelor Thesis Data/results/Distance Analysis/Lasso Regression\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d934c83",
   "metadata": {},
   "source": [
    "## Loading Training and Test Split Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f662c7f",
   "metadata": {},
   "source": [
    "Within this section, we can load the training and test split data files from the directory 'data_directory_training_and_test_splits' into this notebook by calling the function 'pd.read_csv()' with as a parameter the to be read file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99b2d316",
   "metadata": {},
   "source": [
    "#### Loading the 'fold_assignments_samples.csv' file into this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9e36c6b7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:01:00.546425Z",
     "start_time": "2023-06-16T21:01:00.472383Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 'fold_assignments' DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Fold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-06-0125-01A-01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-06-0125-02A-11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-06-0152-02A-01</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-06-0171-02A-11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-06-0190-01A-01</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TCGA-76-4927-01A-01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TCGA-76-4928-01B-01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>TCGA-76-4929-01A-01</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TCGA-76-4931-01A-01</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TCGA-76-4932-01A-01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Samples  Fold\n",
       "0   TCGA-06-0125-01A-01     1\n",
       "1   TCGA-06-0125-02A-11     1\n",
       "2   TCGA-06-0152-02A-01     2\n",
       "3   TCGA-06-0171-02A-11     1\n",
       "4   TCGA-06-0190-01A-01     4\n",
       "..                  ...   ...\n",
       "59  TCGA-76-4927-01A-01     1\n",
       "60  TCGA-76-4928-01B-01     3\n",
       "61  TCGA-76-4929-01A-01     2\n",
       "62  TCGA-76-4931-01A-01     2\n",
       "63  TCGA-76-4932-01A-01     1\n",
       "\n",
       "[64 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the file 'fold_assignments_samples.csv'.\n",
    "fold_assignments = pd.read_csv(data_directory_training_and_test_splits + '/fold_assignments_samples.csv')\n",
    "\n",
    "print(\"The 'fold_assignments' DataFrame:\")\n",
    "fold_assignments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4696519",
   "metadata": {},
   "source": [
    "#### Loading the 'training_and_test_assignments_samples.csv' file into this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "04db2fec",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:01:00.609193Z",
     "start_time": "2023-06-16T21:01:00.551658Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 'training_and_test_assignments' DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Split 1</th>\n",
       "      <th>Split 2</th>\n",
       "      <th>Split 3</th>\n",
       "      <th>Split 4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-06-0125-01A-01</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-06-0125-02A-11</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-06-0152-02A-01</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-06-0171-02A-11</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-06-0190-01A-01</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TEST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TCGA-76-4927-01A-01</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TCGA-76-4928-01B-01</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>TCGA-76-4929-01A-01</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TCGA-76-4931-01A-01</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TCGA-76-4932-01A-01</td>\n",
       "      <td>TEST</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "      <td>TRAIN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Samples Split 1 Split 2 Split 3 Split 4\n",
       "0   TCGA-06-0125-01A-01    TEST   TRAIN   TRAIN   TRAIN\n",
       "1   TCGA-06-0125-02A-11    TEST   TRAIN   TRAIN   TRAIN\n",
       "2   TCGA-06-0152-02A-01   TRAIN    TEST   TRAIN   TRAIN\n",
       "3   TCGA-06-0171-02A-11    TEST   TRAIN   TRAIN   TRAIN\n",
       "4   TCGA-06-0190-01A-01   TRAIN   TRAIN   TRAIN    TEST\n",
       "..                  ...     ...     ...     ...     ...\n",
       "59  TCGA-76-4927-01A-01    TEST   TRAIN   TRAIN   TRAIN\n",
       "60  TCGA-76-4928-01B-01   TRAIN   TRAIN    TEST   TRAIN\n",
       "61  TCGA-76-4929-01A-01   TRAIN    TEST   TRAIN   TRAIN\n",
       "62  TCGA-76-4931-01A-01   TRAIN    TEST   TRAIN   TRAIN\n",
       "63  TCGA-76-4932-01A-01    TEST   TRAIN   TRAIN   TRAIN\n",
       "\n",
       "[64 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the file 'training_and_test_assignments_samples.csv'.\n",
    "training_and_test_assignments = pd.read_csv(data_directory_training_and_test_splits + '/training_and_test_assignments_samples.csv')\n",
    "\n",
    "print(\"The 'training_and_test_assignments' DataFrame:\")\n",
    "training_and_test_assignments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecf15a2e",
   "metadata": {},
   "source": [
    "## Loading all the Different Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4728232",
   "metadata": {},
   "source": [
    "Within this section, the datasets present within the best performing dataset combination are loaded into this notebook:\n",
    "* The M-transformed methylation data file\n",
    "* The log2-transformed gene expression data file\n",
    "\n",
    "These are present in the directories 'data_directory_final_datasets' and 'data_directory_final_datasets_distance'. For each of the corresponding files, this can be achieved by calling the function 'pd.read_csv()' with as a parameter the to be read file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77f6a9fd",
   "metadata": {},
   "source": [
    "#### Loading the 'methylation_data_M_transformed_final.csv' file into this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "39ce7e94",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:06:19.790119Z",
     "start_time": "2023-06-16T21:01:00.615194Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 'methylation_data_M_transformed' DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>cg00000957</th>\n",
       "      <th>cg00001349</th>\n",
       "      <th>cg00001583</th>\n",
       "      <th>cg00002837</th>\n",
       "      <th>cg00003287</th>\n",
       "      <th>cg00004121</th>\n",
       "      <th>cg00008647</th>\n",
       "      <th>cg00009292</th>\n",
       "      <th>cg00011717</th>\n",
       "      <th>...</th>\n",
       "      <th>ch.22.28920330F</th>\n",
       "      <th>ch.22.436090R</th>\n",
       "      <th>ch.22.441164F</th>\n",
       "      <th>ch.22.528917R</th>\n",
       "      <th>ch.22.569473R</th>\n",
       "      <th>ch.22.707049R</th>\n",
       "      <th>ch.22.728807R</th>\n",
       "      <th>ch.22.734399R</th>\n",
       "      <th>ch.22.772318F</th>\n",
       "      <th>ch.22.909671F</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-06-0125-01A-01</td>\n",
       "      <td>3.132755</td>\n",
       "      <td>3.960518</td>\n",
       "      <td>-5.452737</td>\n",
       "      <td>2.348422</td>\n",
       "      <td>0.642046</td>\n",
       "      <td>0.425173</td>\n",
       "      <td>-3.568310</td>\n",
       "      <td>0.090094</td>\n",
       "      <td>4.677447</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.328807</td>\n",
       "      <td>-1.604700</td>\n",
       "      <td>-5.391144</td>\n",
       "      <td>-4.299671</td>\n",
       "      <td>-3.299155</td>\n",
       "      <td>-5.006189</td>\n",
       "      <td>-4.483695</td>\n",
       "      <td>-1.512707</td>\n",
       "      <td>-4.935511</td>\n",
       "      <td>-4.526937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-06-0125-02A-11</td>\n",
       "      <td>3.196057</td>\n",
       "      <td>3.825019</td>\n",
       "      <td>-5.503606</td>\n",
       "      <td>1.372434</td>\n",
       "      <td>0.849407</td>\n",
       "      <td>0.629880</td>\n",
       "      <td>-3.292764</td>\n",
       "      <td>1.242929</td>\n",
       "      <td>5.700119</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.854379</td>\n",
       "      <td>-1.720475</td>\n",
       "      <td>-5.169584</td>\n",
       "      <td>-4.430285</td>\n",
       "      <td>-3.100187</td>\n",
       "      <td>-4.953307</td>\n",
       "      <td>-3.404266</td>\n",
       "      <td>-1.763778</td>\n",
       "      <td>-4.931599</td>\n",
       "      <td>-3.668569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-06-0152-02A-01</td>\n",
       "      <td>4.057813</td>\n",
       "      <td>3.626717</td>\n",
       "      <td>1.146710</td>\n",
       "      <td>-0.208610</td>\n",
       "      <td>0.059986</td>\n",
       "      <td>0.788350</td>\n",
       "      <td>-0.941862</td>\n",
       "      <td>1.416831</td>\n",
       "      <td>5.731835</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.614915</td>\n",
       "      <td>-2.221432</td>\n",
       "      <td>-5.487836</td>\n",
       "      <td>-4.947837</td>\n",
       "      <td>-2.324764</td>\n",
       "      <td>-4.632719</td>\n",
       "      <td>-4.005875</td>\n",
       "      <td>-1.983516</td>\n",
       "      <td>-4.787276</td>\n",
       "      <td>-3.809136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-06-0171-02A-11</td>\n",
       "      <td>4.139295</td>\n",
       "      <td>3.058785</td>\n",
       "      <td>-1.109405</td>\n",
       "      <td>-0.179801</td>\n",
       "      <td>-2.005682</td>\n",
       "      <td>0.634832</td>\n",
       "      <td>-4.202176</td>\n",
       "      <td>-0.933237</td>\n",
       "      <td>6.317184</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.329587</td>\n",
       "      <td>-3.007868</td>\n",
       "      <td>-5.728251</td>\n",
       "      <td>-3.993713</td>\n",
       "      <td>-3.442227</td>\n",
       "      <td>-4.986055</td>\n",
       "      <td>-3.436608</td>\n",
       "      <td>-2.831527</td>\n",
       "      <td>-5.334004</td>\n",
       "      <td>-4.291577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-06-0190-01A-01</td>\n",
       "      <td>3.179215</td>\n",
       "      <td>3.408476</td>\n",
       "      <td>-4.613496</td>\n",
       "      <td>-0.233366</td>\n",
       "      <td>-0.672902</td>\n",
       "      <td>0.370326</td>\n",
       "      <td>-1.536587</td>\n",
       "      <td>0.712060</td>\n",
       "      <td>4.594485</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.925038</td>\n",
       "      <td>-2.666686</td>\n",
       "      <td>-5.455511</td>\n",
       "      <td>-4.393648</td>\n",
       "      <td>-3.584784</td>\n",
       "      <td>-5.394690</td>\n",
       "      <td>-3.741805</td>\n",
       "      <td>-2.292808</td>\n",
       "      <td>-4.840999</td>\n",
       "      <td>-3.959180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TCGA-76-4927-01A-01</td>\n",
       "      <td>3.427987</td>\n",
       "      <td>4.101014</td>\n",
       "      <td>-0.268118</td>\n",
       "      <td>-0.005975</td>\n",
       "      <td>-0.769480</td>\n",
       "      <td>0.898056</td>\n",
       "      <td>-5.020576</td>\n",
       "      <td>1.241059</td>\n",
       "      <td>5.889282</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.408250</td>\n",
       "      <td>-1.532935</td>\n",
       "      <td>-3.325032</td>\n",
       "      <td>-2.215352</td>\n",
       "      <td>-3.453181</td>\n",
       "      <td>-3.190405</td>\n",
       "      <td>-3.143200</td>\n",
       "      <td>-0.768262</td>\n",
       "      <td>-2.643066</td>\n",
       "      <td>-1.345063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TCGA-76-4928-01B-01</td>\n",
       "      <td>2.091628</td>\n",
       "      <td>3.824918</td>\n",
       "      <td>-5.990499</td>\n",
       "      <td>0.344366</td>\n",
       "      <td>-0.717767</td>\n",
       "      <td>0.425782</td>\n",
       "      <td>-5.487022</td>\n",
       "      <td>0.370565</td>\n",
       "      <td>5.714939</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.370823</td>\n",
       "      <td>-1.044793</td>\n",
       "      <td>-4.456032</td>\n",
       "      <td>-2.967873</td>\n",
       "      <td>-3.687673</td>\n",
       "      <td>-3.546754</td>\n",
       "      <td>-3.881002</td>\n",
       "      <td>-1.597628</td>\n",
       "      <td>-4.209691</td>\n",
       "      <td>-2.319970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>TCGA-76-4929-01A-01</td>\n",
       "      <td>3.166884</td>\n",
       "      <td>3.437609</td>\n",
       "      <td>-6.054600</td>\n",
       "      <td>1.700659</td>\n",
       "      <td>-4.722319</td>\n",
       "      <td>0.514961</td>\n",
       "      <td>-5.351739</td>\n",
       "      <td>1.562834</td>\n",
       "      <td>3.495553</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.206406</td>\n",
       "      <td>-1.537983</td>\n",
       "      <td>-3.914480</td>\n",
       "      <td>-2.712960</td>\n",
       "      <td>-3.349122</td>\n",
       "      <td>-2.192265</td>\n",
       "      <td>-2.538937</td>\n",
       "      <td>-1.356304</td>\n",
       "      <td>-2.419188</td>\n",
       "      <td>-1.598524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TCGA-76-4931-01A-01</td>\n",
       "      <td>2.464759</td>\n",
       "      <td>3.631037</td>\n",
       "      <td>3.468339</td>\n",
       "      <td>0.585791</td>\n",
       "      <td>0.666281</td>\n",
       "      <td>0.879107</td>\n",
       "      <td>-5.656808</td>\n",
       "      <td>3.381448</td>\n",
       "      <td>4.041543</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.784318</td>\n",
       "      <td>-1.500444</td>\n",
       "      <td>-4.241542</td>\n",
       "      <td>-4.235881</td>\n",
       "      <td>-3.584725</td>\n",
       "      <td>-2.905236</td>\n",
       "      <td>-3.847968</td>\n",
       "      <td>-0.846516</td>\n",
       "      <td>-4.245142</td>\n",
       "      <td>-1.847410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TCGA-76-4932-01A-01</td>\n",
       "      <td>3.020878</td>\n",
       "      <td>3.532181</td>\n",
       "      <td>-2.759466</td>\n",
       "      <td>3.165462</td>\n",
       "      <td>-4.026015</td>\n",
       "      <td>0.813059</td>\n",
       "      <td>-5.397459</td>\n",
       "      <td>2.687827</td>\n",
       "      <td>5.128151</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.484844</td>\n",
       "      <td>-1.510911</td>\n",
       "      <td>-3.753825</td>\n",
       "      <td>-3.266255</td>\n",
       "      <td>-3.624682</td>\n",
       "      <td>-3.464562</td>\n",
       "      <td>-3.156996</td>\n",
       "      <td>-1.384458</td>\n",
       "      <td>-3.694331</td>\n",
       "      <td>-2.024216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 270879 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Samples  cg00000957  cg00001349  cg00001583  cg00002837  \\\n",
       "0   TCGA-06-0125-01A-01    3.132755    3.960518   -5.452737    2.348422   \n",
       "1   TCGA-06-0125-02A-11    3.196057    3.825019   -5.503606    1.372434   \n",
       "2   TCGA-06-0152-02A-01    4.057813    3.626717    1.146710   -0.208610   \n",
       "3   TCGA-06-0171-02A-11    4.139295    3.058785   -1.109405   -0.179801   \n",
       "4   TCGA-06-0190-01A-01    3.179215    3.408476   -4.613496   -0.233366   \n",
       "..                  ...         ...         ...         ...         ...   \n",
       "59  TCGA-76-4927-01A-01    3.427987    4.101014   -0.268118   -0.005975   \n",
       "60  TCGA-76-4928-01B-01    2.091628    3.824918   -5.990499    0.344366   \n",
       "61  TCGA-76-4929-01A-01    3.166884    3.437609   -6.054600    1.700659   \n",
       "62  TCGA-76-4931-01A-01    2.464759    3.631037    3.468339    0.585791   \n",
       "63  TCGA-76-4932-01A-01    3.020878    3.532181   -2.759466    3.165462   \n",
       "\n",
       "    cg00003287  cg00004121  cg00008647  cg00009292  cg00011717  ...  \\\n",
       "0     0.642046    0.425173   -3.568310    0.090094    4.677447  ...   \n",
       "1     0.849407    0.629880   -3.292764    1.242929    5.700119  ...   \n",
       "2     0.059986    0.788350   -0.941862    1.416831    5.731835  ...   \n",
       "3    -2.005682    0.634832   -4.202176   -0.933237    6.317184  ...   \n",
       "4    -0.672902    0.370326   -1.536587    0.712060    4.594485  ...   \n",
       "..         ...         ...         ...         ...         ...  ...   \n",
       "59   -0.769480    0.898056   -5.020576    1.241059    5.889282  ...   \n",
       "60   -0.717767    0.425782   -5.487022    0.370565    5.714939  ...   \n",
       "61   -4.722319    0.514961   -5.351739    1.562834    3.495553  ...   \n",
       "62    0.666281    0.879107   -5.656808    3.381448    4.041543  ...   \n",
       "63   -4.026015    0.813059   -5.397459    2.687827    5.128151  ...   \n",
       "\n",
       "    ch.22.28920330F  ch.22.436090R  ch.22.441164F  ch.22.528917R  \\\n",
       "0         -4.328807      -1.604700      -5.391144      -4.299671   \n",
       "1         -2.854379      -1.720475      -5.169584      -4.430285   \n",
       "2         -4.614915      -2.221432      -5.487836      -4.947837   \n",
       "3         -3.329587      -3.007868      -5.728251      -3.993713   \n",
       "4         -3.925038      -2.666686      -5.455511      -4.393648   \n",
       "..              ...            ...            ...            ...   \n",
       "59        -4.408250      -1.532935      -3.325032      -2.215352   \n",
       "60        -5.370823      -1.044793      -4.456032      -2.967873   \n",
       "61        -4.206406      -1.537983      -3.914480      -2.712960   \n",
       "62        -4.784318      -1.500444      -4.241542      -4.235881   \n",
       "63        -3.484844      -1.510911      -3.753825      -3.266255   \n",
       "\n",
       "    ch.22.569473R  ch.22.707049R  ch.22.728807R  ch.22.734399R  ch.22.772318F  \\\n",
       "0       -3.299155      -5.006189      -4.483695      -1.512707      -4.935511   \n",
       "1       -3.100187      -4.953307      -3.404266      -1.763778      -4.931599   \n",
       "2       -2.324764      -4.632719      -4.005875      -1.983516      -4.787276   \n",
       "3       -3.442227      -4.986055      -3.436608      -2.831527      -5.334004   \n",
       "4       -3.584784      -5.394690      -3.741805      -2.292808      -4.840999   \n",
       "..            ...            ...            ...            ...            ...   \n",
       "59      -3.453181      -3.190405      -3.143200      -0.768262      -2.643066   \n",
       "60      -3.687673      -3.546754      -3.881002      -1.597628      -4.209691   \n",
       "61      -3.349122      -2.192265      -2.538937      -1.356304      -2.419188   \n",
       "62      -3.584725      -2.905236      -3.847968      -0.846516      -4.245142   \n",
       "63      -3.624682      -3.464562      -3.156996      -1.384458      -3.694331   \n",
       "\n",
       "    ch.22.909671F  \n",
       "0       -4.526937  \n",
       "1       -3.668569  \n",
       "2       -3.809136  \n",
       "3       -4.291577  \n",
       "4       -3.959180  \n",
       "..            ...  \n",
       "59      -1.345063  \n",
       "60      -2.319970  \n",
       "61      -1.598524  \n",
       "62      -1.847410  \n",
       "63      -2.024216  \n",
       "\n",
       "[64 rows x 270879 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the file 'methylation_data_M_transformed_final.csv'.\n",
    "methylation_data_M_transformed = pd.read_csv(data_directory_final_datasets + '/methylation_data_M_transformed_final.csv')\n",
    "\n",
    "print(\"The 'methylation_data_M_transformed' DataFrame:\")\n",
    "methylation_data_M_transformed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99337900",
   "metadata": {},
   "source": [
    "#### Loading the 'gene_expression_data_log2_transformed_distance_correlation_genes_removed.csv' file into this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d65a5761",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:06:20.537178Z",
     "start_time": "2023-06-16T21:06:19.793130Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 'gene_expression_data_log2_transformed' DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>ENSG00000001561</th>\n",
       "      <th>ENSG00000001629</th>\n",
       "      <th>ENSG00000001631</th>\n",
       "      <th>ENSG00000002587</th>\n",
       "      <th>ENSG00000002746</th>\n",
       "      <th>ENSG00000004487</th>\n",
       "      <th>ENSG00000004534</th>\n",
       "      <th>ENSG00000004777</th>\n",
       "      <th>ENSG00000005007</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000287064</th>\n",
       "      <th>ENSG00000287151</th>\n",
       "      <th>ENSG00000287263</th>\n",
       "      <th>ENSG00000287562</th>\n",
       "      <th>ENSG00000287828</th>\n",
       "      <th>ENSG00000287893</th>\n",
       "      <th>ENSG00000288156</th>\n",
       "      <th>ENSG00000288586</th>\n",
       "      <th>ENSG00000288612</th>\n",
       "      <th>ENSG00000288658</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-06-0125-01A-01</td>\n",
       "      <td>1.800910</td>\n",
       "      <td>6.291825</td>\n",
       "      <td>3.187578</td>\n",
       "      <td>4.739843</td>\n",
       "      <td>0.840443</td>\n",
       "      <td>6.443154</td>\n",
       "      <td>6.150584</td>\n",
       "      <td>5.753300</td>\n",
       "      <td>6.185981</td>\n",
       "      <td>...</td>\n",
       "      <td>0.383055</td>\n",
       "      <td>0.998123</td>\n",
       "      <td>2.884598</td>\n",
       "      <td>2.785362</td>\n",
       "      <td>1.155102</td>\n",
       "      <td>0.662205</td>\n",
       "      <td>1.714795</td>\n",
       "      <td>1.547549</td>\n",
       "      <td>2.704385</td>\n",
       "      <td>0.767655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-06-0125-02A-11</td>\n",
       "      <td>3.646969</td>\n",
       "      <td>6.455447</td>\n",
       "      <td>2.624943</td>\n",
       "      <td>3.701738</td>\n",
       "      <td>1.102591</td>\n",
       "      <td>5.885086</td>\n",
       "      <td>4.452912</td>\n",
       "      <td>4.893236</td>\n",
       "      <td>6.229463</td>\n",
       "      <td>...</td>\n",
       "      <td>1.371670</td>\n",
       "      <td>0.805292</td>\n",
       "      <td>2.562279</td>\n",
       "      <td>1.262373</td>\n",
       "      <td>0.684101</td>\n",
       "      <td>1.156397</td>\n",
       "      <td>2.349592</td>\n",
       "      <td>1.382944</td>\n",
       "      <td>1.640852</td>\n",
       "      <td>0.700617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-06-0152-02A-01</td>\n",
       "      <td>3.470029</td>\n",
       "      <td>6.264604</td>\n",
       "      <td>2.771886</td>\n",
       "      <td>2.792501</td>\n",
       "      <td>1.866354</td>\n",
       "      <td>6.251889</td>\n",
       "      <td>5.459441</td>\n",
       "      <td>5.540536</td>\n",
       "      <td>6.376085</td>\n",
       "      <td>...</td>\n",
       "      <td>2.235666</td>\n",
       "      <td>1.427284</td>\n",
       "      <td>1.817582</td>\n",
       "      <td>1.778503</td>\n",
       "      <td>2.435389</td>\n",
       "      <td>0.268075</td>\n",
       "      <td>2.293047</td>\n",
       "      <td>1.481764</td>\n",
       "      <td>2.196733</td>\n",
       "      <td>1.697285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-06-0171-02A-11</td>\n",
       "      <td>5.374226</td>\n",
       "      <td>5.630117</td>\n",
       "      <td>2.075122</td>\n",
       "      <td>2.951569</td>\n",
       "      <td>0.387693</td>\n",
       "      <td>4.875726</td>\n",
       "      <td>4.920541</td>\n",
       "      <td>3.527258</td>\n",
       "      <td>5.813684</td>\n",
       "      <td>...</td>\n",
       "      <td>0.524866</td>\n",
       "      <td>1.275186</td>\n",
       "      <td>3.038612</td>\n",
       "      <td>1.359971</td>\n",
       "      <td>0.718526</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.087887</td>\n",
       "      <td>0.931078</td>\n",
       "      <td>2.126444</td>\n",
       "      <td>0.766722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-06-0190-01A-01</td>\n",
       "      <td>4.371350</td>\n",
       "      <td>5.433400</td>\n",
       "      <td>2.516519</td>\n",
       "      <td>2.700994</td>\n",
       "      <td>0.483777</td>\n",
       "      <td>6.042990</td>\n",
       "      <td>5.195911</td>\n",
       "      <td>5.074073</td>\n",
       "      <td>5.765248</td>\n",
       "      <td>...</td>\n",
       "      <td>0.504570</td>\n",
       "      <td>1.549620</td>\n",
       "      <td>3.181452</td>\n",
       "      <td>1.369606</td>\n",
       "      <td>2.538563</td>\n",
       "      <td>0.603976</td>\n",
       "      <td>3.092427</td>\n",
       "      <td>0.928048</td>\n",
       "      <td>1.875387</td>\n",
       "      <td>0.229834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TCGA-76-4927-01A-01</td>\n",
       "      <td>3.829444</td>\n",
       "      <td>6.280969</td>\n",
       "      <td>2.682978</td>\n",
       "      <td>3.194339</td>\n",
       "      <td>1.531768</td>\n",
       "      <td>6.083558</td>\n",
       "      <td>5.055677</td>\n",
       "      <td>5.403540</td>\n",
       "      <td>6.111363</td>\n",
       "      <td>...</td>\n",
       "      <td>1.157367</td>\n",
       "      <td>2.028640</td>\n",
       "      <td>3.703887</td>\n",
       "      <td>1.156526</td>\n",
       "      <td>1.974860</td>\n",
       "      <td>1.064469</td>\n",
       "      <td>2.707591</td>\n",
       "      <td>1.171783</td>\n",
       "      <td>1.187134</td>\n",
       "      <td>1.310689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TCGA-76-4928-01B-01</td>\n",
       "      <td>3.613025</td>\n",
       "      <td>5.277862</td>\n",
       "      <td>2.277092</td>\n",
       "      <td>2.464747</td>\n",
       "      <td>0.854634</td>\n",
       "      <td>6.027249</td>\n",
       "      <td>4.900635</td>\n",
       "      <td>4.202339</td>\n",
       "      <td>5.355739</td>\n",
       "      <td>...</td>\n",
       "      <td>0.431035</td>\n",
       "      <td>1.828022</td>\n",
       "      <td>2.732334</td>\n",
       "      <td>1.098824</td>\n",
       "      <td>1.462733</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.435789</td>\n",
       "      <td>0.440952</td>\n",
       "      <td>1.622790</td>\n",
       "      <td>0.516923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>TCGA-76-4929-01A-01</td>\n",
       "      <td>4.978159</td>\n",
       "      <td>6.151981</td>\n",
       "      <td>2.740604</td>\n",
       "      <td>2.949591</td>\n",
       "      <td>0.632827</td>\n",
       "      <td>7.047989</td>\n",
       "      <td>5.192352</td>\n",
       "      <td>4.957706</td>\n",
       "      <td>5.554177</td>\n",
       "      <td>...</td>\n",
       "      <td>0.276556</td>\n",
       "      <td>2.139044</td>\n",
       "      <td>4.254881</td>\n",
       "      <td>2.153546</td>\n",
       "      <td>1.410287</td>\n",
       "      <td>2.314174</td>\n",
       "      <td>0.731965</td>\n",
       "      <td>2.031395</td>\n",
       "      <td>1.956837</td>\n",
       "      <td>2.229649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TCGA-76-4931-01A-01</td>\n",
       "      <td>3.653003</td>\n",
       "      <td>6.518490</td>\n",
       "      <td>3.018225</td>\n",
       "      <td>3.815545</td>\n",
       "      <td>1.211884</td>\n",
       "      <td>6.588232</td>\n",
       "      <td>6.448640</td>\n",
       "      <td>6.178647</td>\n",
       "      <td>6.011404</td>\n",
       "      <td>...</td>\n",
       "      <td>1.364012</td>\n",
       "      <td>2.444163</td>\n",
       "      <td>2.508403</td>\n",
       "      <td>2.561839</td>\n",
       "      <td>2.329698</td>\n",
       "      <td>1.014141</td>\n",
       "      <td>1.270589</td>\n",
       "      <td>1.569199</td>\n",
       "      <td>3.349351</td>\n",
       "      <td>0.928579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TCGA-76-4932-01A-01</td>\n",
       "      <td>1.820935</td>\n",
       "      <td>5.882095</td>\n",
       "      <td>2.230142</td>\n",
       "      <td>4.009831</td>\n",
       "      <td>0.167358</td>\n",
       "      <td>5.385417</td>\n",
       "      <td>4.720268</td>\n",
       "      <td>5.228572</td>\n",
       "      <td>5.626921</td>\n",
       "      <td>...</td>\n",
       "      <td>0.172872</td>\n",
       "      <td>1.681944</td>\n",
       "      <td>1.929450</td>\n",
       "      <td>1.250840</td>\n",
       "      <td>1.600841</td>\n",
       "      <td>0.746227</td>\n",
       "      <td>2.089057</td>\n",
       "      <td>0.902421</td>\n",
       "      <td>1.092410</td>\n",
       "      <td>1.557288</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 3444 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Samples  ENSG00000001561  ENSG00000001629  ENSG00000001631  \\\n",
       "0   TCGA-06-0125-01A-01         1.800910         6.291825         3.187578   \n",
       "1   TCGA-06-0125-02A-11         3.646969         6.455447         2.624943   \n",
       "2   TCGA-06-0152-02A-01         3.470029         6.264604         2.771886   \n",
       "3   TCGA-06-0171-02A-11         5.374226         5.630117         2.075122   \n",
       "4   TCGA-06-0190-01A-01         4.371350         5.433400         2.516519   \n",
       "..                  ...              ...              ...              ...   \n",
       "59  TCGA-76-4927-01A-01         3.829444         6.280969         2.682978   \n",
       "60  TCGA-76-4928-01B-01         3.613025         5.277862         2.277092   \n",
       "61  TCGA-76-4929-01A-01         4.978159         6.151981         2.740604   \n",
       "62  TCGA-76-4931-01A-01         3.653003         6.518490         3.018225   \n",
       "63  TCGA-76-4932-01A-01         1.820935         5.882095         2.230142   \n",
       "\n",
       "    ENSG00000002587  ENSG00000002746  ENSG00000004487  ENSG00000004534  \\\n",
       "0          4.739843         0.840443         6.443154         6.150584   \n",
       "1          3.701738         1.102591         5.885086         4.452912   \n",
       "2          2.792501         1.866354         6.251889         5.459441   \n",
       "3          2.951569         0.387693         4.875726         4.920541   \n",
       "4          2.700994         0.483777         6.042990         5.195911   \n",
       "..              ...              ...              ...              ...   \n",
       "59         3.194339         1.531768         6.083558         5.055677   \n",
       "60         2.464747         0.854634         6.027249         4.900635   \n",
       "61         2.949591         0.632827         7.047989         5.192352   \n",
       "62         3.815545         1.211884         6.588232         6.448640   \n",
       "63         4.009831         0.167358         5.385417         4.720268   \n",
       "\n",
       "    ENSG00000004777  ENSG00000005007  ...  ENSG00000287064  ENSG00000287151  \\\n",
       "0          5.753300         6.185981  ...         0.383055         0.998123   \n",
       "1          4.893236         6.229463  ...         1.371670         0.805292   \n",
       "2          5.540536         6.376085  ...         2.235666         1.427284   \n",
       "3          3.527258         5.813684  ...         0.524866         1.275186   \n",
       "4          5.074073         5.765248  ...         0.504570         1.549620   \n",
       "..              ...              ...  ...              ...              ...   \n",
       "59         5.403540         6.111363  ...         1.157367         2.028640   \n",
       "60         4.202339         5.355739  ...         0.431035         1.828022   \n",
       "61         4.957706         5.554177  ...         0.276556         2.139044   \n",
       "62         6.178647         6.011404  ...         1.364012         2.444163   \n",
       "63         5.228572         5.626921  ...         0.172872         1.681944   \n",
       "\n",
       "    ENSG00000287263  ENSG00000287562  ENSG00000287828  ENSG00000287893  \\\n",
       "0          2.884598         2.785362         1.155102         0.662205   \n",
       "1          2.562279         1.262373         0.684101         1.156397   \n",
       "2          1.817582         1.778503         2.435389         0.268075   \n",
       "3          3.038612         1.359971         0.718526         0.000000   \n",
       "4          3.181452         1.369606         2.538563         0.603976   \n",
       "..              ...              ...              ...              ...   \n",
       "59         3.703887         1.156526         1.974860         1.064469   \n",
       "60         2.732334         1.098824         1.462733         0.000000   \n",
       "61         4.254881         2.153546         1.410287         2.314174   \n",
       "62         2.508403         2.561839         2.329698         1.014141   \n",
       "63         1.929450         1.250840         1.600841         0.746227   \n",
       "\n",
       "    ENSG00000288156  ENSG00000288586  ENSG00000288612  ENSG00000288658  \n",
       "0          1.714795         1.547549         2.704385         0.767655  \n",
       "1          2.349592         1.382944         1.640852         0.700617  \n",
       "2          2.293047         1.481764         2.196733         1.697285  \n",
       "3          4.087887         0.931078         2.126444         0.766722  \n",
       "4          3.092427         0.928048         1.875387         0.229834  \n",
       "..              ...              ...              ...              ...  \n",
       "59         2.707591         1.171783         1.187134         1.310689  \n",
       "60         2.435789         0.440952         1.622790         0.516923  \n",
       "61         0.731965         2.031395         1.956837         2.229649  \n",
       "62         1.270589         1.569199         3.349351         0.928579  \n",
       "63         2.089057         0.902421         1.092410         1.557288  \n",
       "\n",
       "[64 rows x 3444 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the file 'gene_expression_data_log2_transformed_distance_correlation_genes_removed.csv'.\n",
    "gene_expression_data_log2_transformed = pd.read_csv(data_directory_final_datasets_distance + '/gene_expression_data_log2_transformed_distance_correlation_genes_removed.csv')\n",
    "\n",
    "print(\"The 'gene_expression_data_log2_transformed' DataFrame:\")\n",
    "gene_expression_data_log2_transformed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "175440ae",
   "metadata": {},
   "source": [
    "## Lasso Regression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "202850ea",
   "metadata": {},
   "source": [
    "Within this section, lasso regression is applied for the dataset combination loaded into this notebook above which is performed multiple times, once for each distance defined below:\n",
    "* 5,000 \n",
    "* 10,000\n",
    "* 15,000\n",
    "* 25,000\n",
    "* 50,000\n",
    "* 75,000\n",
    "* 100,000\n",
    "* 150,000\n",
    "* 250,000\n",
    "* 350,000\n",
    "* 500,000\n",
    "* 750,000\n",
    "* 1,000,000\n",
    "* 1,500,000\n",
    "* 2,000,000\n",
    "* 2,500,000\n",
    "* 4,000,000\n",
    "* 5,000,000\n",
    "* 6,000,000\n",
    "* 7,500,000\n",
    "* 10,000,000\n",
    "* 12,500,000\n",
    "* 15,000,000\n",
    "* 17,500,000\n",
    "* 20,000,000\n",
    "* 25,000,000\n",
    "* 30,000,000\n",
    "* 40,000,000\n",
    "* 50,000,000\n",
    "* 65,000,000\n",
    "* 80,000,000\n",
    "* 100,000,000\n",
    "* 120,000,000\n",
    "* 150,000,000\n",
    "* 200,000,000\n",
    "* 250,000,000\n",
    "* 350,000,000\n",
    "* 500,000,000\n",
    "* 750,000,000\n",
    "* 1,000,000,000\n",
    "\n",
    "For each of these distances lasso regression models are built, one for each gene, and these are evaluated. The distance of the best performing experiment will represent which distance we should use to determine which CpG sites to consider (based on their position relative to a particular gene) for predicting the gene expression value of a particular gene.\n",
    "\n",
    "As mentioned within the 'Introduction', we will use all of the genes present within the log2-transformed gene expression data, and not just the ones present on a single chromosome, since this notebook (among the other machine learning algorithms notebooks within this directory) represents the main experiments of the Distance Analysis part.\n",
    "\n",
    "To retrieve the prediction accuracy of the gene expression values, the R-squared (R<sup>2</sup>) metric is computed for each of the predictions which indicates the proportion of the variance in the dependent variable that can be explained by the model [3]. Higher R<sup>2</sup> values indicate a more significant proportion of the variance in the dependent variable that can be explained by the model, with 1 being the largest possible value. This R<sup>2</sup> value is retrieved by applying 4-fold cross-validation using the training and test splits defined in the notebook 'Training and Test Set Division.ipynb' present in the 'Machine Learning Algorithms - Preprocessing' folder, which also includes the motivation behind choosing the k in k-fold cross-validation to be set equal to four, and averaging the R<sup>2</sup> value for each of the four folds. The resulting R<sup>2</sup> values for each of the genes for each of the distances listed above will be visualized within a single box plot in the notebook 'Analyzing Lasso Regression Results for Distance Analysis'."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33e9dc41",
   "metadata": {},
   "source": [
    "The first thing we can do is run the 'Machine Learning Additional Functions.ipynb' notebook present in the folder 'Machine Learning Algorithms' which contains additional helper functions, such as retrieving the methylation data within a certain distance from a gene, for the machine learning algorithms. This notebook can be run by calling the command '%run' with as argument the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "431ada1c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:06:21.320314Z",
     "start_time": "2023-06-16T21:06:20.542176Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting the importing of the libraries...\n",
      "Finishing the installing of the libraries.\n",
      "The library 'pd' (pandas) has been loaded into the notebook with its version being:\n",
      "1.2.4\n",
      "The library 'np' (numpy) has been loaded into the notebook with its version being:\n",
      "1.20.1\n",
      "\n",
      "The library 're' has been loaded into the notebook with its version being:\n",
      "2.2.1\n",
      "\n",
      "The library 'sys' has been loaded into the notebook with its version being:\n",
      "3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]\n",
      "The 'CpG_sites_location_data' DataFrame containing the location data of the CpG sites:\n",
      "The 'genes_location_data' DataFrame containing the location data of the genes:\n",
      "The 'chromosomes_length_data' DataFrame containing the lengths of the chromosomes:\n"
     ]
    }
   ],
   "source": [
    "# Running the notebook 'Machine Learning Additional Functions.ipynb' by calling the command '%run'.\n",
    "%run \"../Machine Learning Additional Functions.ipynb\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0dc7c04",
   "metadata": {},
   "source": [
    "As opposed to the linear regression algorithm where no additional parameters need to be considered, lasso regression has an additional parameter which represents a constant that multiplies the L1 term controlling the lasso regularization strength (the regularization parameter $\\lambda$ presented in the introduction). This constant is commonly referred to as 'alpha' which is the terminology that will also be used throughout the remainder of this notebook. Since this alpha can take any value between 0 and infinity (although 0 is not recommended as this essentially represents the same OLS estimator), it is crucial that we first experiment with different values of alpha for our data to find the one that produces the highest R<sup>2</sup> scores. To achieve this, we can retrieve the R<sup>2</sup> scores by using different values of alpha where we can reduce the computational burden by only building models for the genes which are located on chromosome 1 and select the CpG sites which are present within a distance of 5,000,000 in both directions from the gene. This allows us to retrieve a single R<sup>2</sup> score for each gene for a certain alpha."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5c5212cd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:06:21.724713Z",
     "start_time": "2023-06-16T21:06:21.323704Z"
    }
   },
   "outputs": [],
   "source": [
    "# Retrieving the gene expression data present in the 'gene_expression_data_log2_transformed' DataFrame of which the genes \n",
    "# are present on chromosome 1 by calling the function 'get_gene_expression_data_from_chromosome()' present within the \n",
    "# notebook 'Machine Learning Additional Functions.ipynb'.\n",
    "gene_expression_data_log2_transformed_chromosome1 = get_gene_expression_data_from_chromosome(gene_expression_data_log2_transformed, 1)\n",
    "\n",
    "# Defining the distance setting to be used for the experiment.\n",
    "distance_experiment = 5000000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5a664f9",
   "metadata": {},
   "source": [
    "Before we perform the experiment, we should first define the function that will perform the lasso regression on the data for the experiment of finding the best alpha."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c50e9fb",
   "metadata": {},
   "source": [
    "#### Defining the functions needed to calculate the R<sup>2</sup> scores for a given distance and alpha"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec3f47cb",
   "metadata": {},
   "source": [
    "Next, we can define the functions needed to calculate the R<sup>2</sup> scores for a given distance and alpha. The function 'calculate_R2_scores()' can also be used within the section 'Applying Lasso Regression' later in this notebook. We can utilize the 'joblib' library to parallelize our code as the computations for a single gene do not influence the computations of any other gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8cb4f331",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:06:21.756750Z",
     "start_time": "2023-06-16T21:06:21.727965Z"
    }
   },
   "outputs": [],
   "source": [
    "# This function calculates the R2 scores for a single 'gene' which is present in the 'gene_expression_data_log2_transformed'\n",
    "# DataFrame.\n",
    "def calculate_R2_scores(gene, distance, alpha): \n",
    "    \n",
    "    # Retrieving the log2-transformed gene expression data of the current 'gene' and retrieving the M-transformed \n",
    "    # methylation data that is within a distance of 'distance' from this gene and for which each of the CpG sites has a \n",
    "    # correlation higher than the threshold of 0.3 with the gene (from the notebook 'Machine Learning Preliminary Analysis \n",
    "    # for Distance Analysis - Pearson's Correlation Coefficient.ipynb') by calling the function \n",
    "    # 'get_methylation_data_close_to_gene_and_with_higher_correlation_than_threshold()' which is present in the notebook \n",
    "    # 'Machine Learning Additional Functions.ipynb'.\n",
    "    gene_expression_data_current_gene = gene_expression_data_log2_transformed[['Samples', gene]]\n",
    "    methylation_data_close_to_current_gene = get_methylation_data_close_to_gene_and_with_higher_correlation_than_threshold(methylation_data_M_transformed, gene_expression_data_current_gene, distance=distance, threshold=0.30)\n",
    "\n",
    "    # Defining a list where all the R2 scores for the current gene will be stored. Since one model is built per fold (split),\n",
    "    # the list 'R2_scores_current_gene' will eventually contain 4 elements (4 R2 scores).\n",
    "    R2_scores_current_gene = []\n",
    "    \n",
    "    # Looping over every column in the 'training_and_test_assignments' DataFrame such that 4-fold cross-validation is \n",
    "    # performed using the training and test sets defined within that 'training_and_test_assignments' DataFrame.\n",
    "    for split in training_and_test_assignments.columns[1:]:\n",
    "\n",
    "        # Retrieving the samples which below to the training and test set for the current split 'split'.\n",
    "        selected_samples_train = training_and_test_assignments.loc[training_and_test_assignments[split] == \"TRAIN\", 'Samples'].tolist()\n",
    "        selected_samples_test = training_and_test_assignments.loc[training_and_test_assignments[split] == \"TEST\", 'Samples'].tolist()\n",
    "\n",
    "        # Retrieving the gene expression and methylation data of which the samples belong to the training set.\n",
    "        gene_expression_data_current_gene_train = gene_expression_data_current_gene.loc[gene_expression_data_current_gene['Samples'].isin(selected_samples_train)].drop(columns=['Samples'])\n",
    "        methylation_data_close_to_current_gene_train = methylation_data_close_to_current_gene.loc[methylation_data_close_to_current_gene['Samples'].isin(selected_samples_train)].drop(columns=['Samples'])\n",
    "\n",
    "        # Retrieving the gene expression and methylation data of which the samples belong to the test set.\n",
    "        gene_expression_data_current_gene_test = gene_expression_data_current_gene.loc[gene_expression_data_current_gene['Samples'].isin(selected_samples_test)].drop(columns=['Samples'])\n",
    "        methylation_data_close_to_current_gene_test = methylation_data_close_to_current_gene.loc[methylation_data_close_to_current_gene['Samples'].isin(selected_samples_test)].drop(columns=['Samples'])\n",
    "        \n",
    "        # Creating a new Lasso Regression model by calling the constructor 'Lasso()' with as parameter the constant that \n",
    "        # multiplies the L1 term and calling the function 'fit()' to train the model with as X-data \n",
    "        # 'methylation_data_close_to_current_gene_train' and as Y-data 'gene_expression_data_current_gene_train'.\n",
    "        model_current_gene = Lasso(alpha=alpha, max_iter=2000) \n",
    "        model_current_gene.fit(methylation_data_close_to_current_gene_train, gene_expression_data_current_gene_train)\n",
    "\n",
    "        # Predicting the gene expression values based on the 'methylation_data_close_to_current_gene_test' by calling the \n",
    "        # function 'predict()'.\n",
    "        gene_expression_data_current_gene_predict = model_current_gene.predict(methylation_data_close_to_current_gene_test)\n",
    "        \n",
    "        # Calculating the R2 score by calling the function 'r2_score()' with the actual values \n",
    "        # 'gene_expression_data_current_gene_test' and the predicted values 'gene_expression_data_current_gene_predict'.\n",
    "        R2_score = r2_score(gene_expression_data_current_gene_test, gene_expression_data_current_gene_predict)\n",
    "        \n",
    "        # Adding the 'R2_score' value to the 'R2_scores_current_gene' list by calling the function 'append()'.\n",
    "        R2_scores_current_gene.append(R2_score)\n",
    "\n",
    "    return {'gene': gene, 'R2': np.mean(R2_scores_current_gene)}\n",
    "\n",
    "\n",
    "# This function retrieves the R2 scores for the lasso regression models (one for each gene) fitted to predict the   \n",
    "# 'gene_expression_data_log2_transformed_chromosome1' based on the 'methylation_data_M_transformed'.\n",
    "def lasso_regression_with_alpha_experiment(alpha):\n",
    "    \n",
    "    # Retrieving the genes present in the DataFrame 'gene_expression_data_log2_transformed_chromosome1'.\n",
    "    genes = gene_expression_data_log2_transformed_chromosome1.columns[1:]\n",
    "    \n",
    "    # Defining a list where all the R2 scores (one for each gene) will be stored such that we can later represent these\n",
    "    # within a box plot to compare them with the R2 scores for the other experiments. This can be achieved by calling the\n",
    "    # function 'calculate_R2_scores()' for each of the genes. Since the computations for a single gene do not influence the \n",
    "    # computations of any other gene, we can parallelize the execution of this function by calling the function 'Parallel()' \n",
    "    # from the 'joblib' library.\n",
    "    R2_scores = Parallel(n_jobs=512)(delayed(calculate_R2_scores)(gene, distance_experiment, alpha) for gene in genes)\n",
    "    \n",
    "    # Combining all the key-value pairs into a single dictionary.\n",
    "    R2_scores_dictionary = {R2_score['gene']: R2_score['R2'] for R2_score in R2_scores}\n",
    "    \n",
    "    return R2_scores_dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a11e756",
   "metadata": {},
   "source": [
    "#### Running the experiment to find the best alpha"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96b3ee6f",
   "metadata": {},
   "source": [
    "Next, we can define a list of alphas for which each the function 'lasso_regression_with_alpha_experiment()' is called. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab14824b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T21:06:21.835378Z",
     "start_time": "2023-06-16T21:06:21.760477Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The empty 'R2_0_to_10_df' DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>0.05</th>\n",
       "      <th>0.10</th>\n",
       "      <th>0.15</th>\n",
       "      <th>0.20</th>\n",
       "      <th>0.25</th>\n",
       "      <th>0.30</th>\n",
       "      <th>0.40</th>\n",
       "      <th>0.50</th>\n",
       "      <th>0.75</th>\n",
       "      <th>...</th>\n",
       "      <th>1.50</th>\n",
       "      <th>2.00</th>\n",
       "      <th>3.00</th>\n",
       "      <th>4.00</th>\n",
       "      <th>5.00</th>\n",
       "      <th>6.00</th>\n",
       "      <th>7.00</th>\n",
       "      <th>8.00</th>\n",
       "      <th>9.00</th>\n",
       "      <th>10.00</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004487</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000007341</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000007923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000007933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000010803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>ENSG00000285867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>ENSG00000286383</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>ENSG00000286619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>ENSG00000287064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>ENSG00000287828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>379 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene  0.05  0.10  0.15  0.20  0.25  0.30  0.40  0.50  0.75  \\\n",
       "0    ENSG00000004487   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "1    ENSG00000007341   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "2    ENSG00000007923   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "3    ENSG00000007933   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "4    ENSG00000010803   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "..               ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "374  ENSG00000285867   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "375  ENSG00000286383   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "376  ENSG00000286619   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "377  ENSG00000287064   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "378  ENSG00000287828   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   \n",
       "\n",
       "     ...  1.50  2.00  3.00  4.00  5.00  6.00  7.00  8.00  9.00  10.00  \n",
       "0    ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "1    ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "2    ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "3    ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "4    ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "..   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
       "374  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "375  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "376  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "377  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "378  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN  \n",
       "\n",
       "[379 rows x 21 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The list of alphas for which each the function 'lasso_regression_with_alpha_experiment()' is called. Since alpha can go \n",
    "# till infinity, we can first look at values of 0.10 through 10. \n",
    "alphas = [0.10, 0.15, 0.20, 0.25, 0.30, 0.40, 0.50, 0.75, 1.00, 1.50, 2.00, 3.00, 4.00, 5.00, 6.00, 7.00, 8.00, 9.00, 10.00]\n",
    "\n",
    "# Creating a dictionary which will later contain the lists of R2 scores for the alpha values defined in the list above.\n",
    "alphas_0_to_10 = ['0.10', '0.15', '0.20', '0.25', '0.30', '0.40', '0.50', '0.75', '1.00', '1.50', '2.00', '3.00', '4.00', '5.00', '6.00', '7.00', '8.00', '9.00', '10.00']\n",
    "R2_0_to_10 = {alpha: [] for alpha in alphas_0_to_10}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_0_to_10_df = pd.DataFrame(R2_0_to_10)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_0_to_10_df.insert(0, 'Gene', gene_expression_data_log2_transformed_chromosome1.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_0_to_10_df' DataFrame:\")\n",
    "R2_0_to_10_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab11a7f4",
   "metadata": {},
   "source": [
    "Now, we can call the function 'lasso_regression_with_alpha_experiment()' for each of the alphas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a53fac19",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T20:22:57.025597Z",
     "start_time": "2023-06-16T20:07:41.051022Z"
    }
   },
   "outputs": [],
   "source": [
    "# Looping over every 'alpha' within the 'alphas' list and applying lasso regression to the M-transformed methylation data \n",
    "# and log2-transformed gene expression data for the 'alpha' and adding their resulting lists filled with R2 scores to the \n",
    "# 'R2_0_to_10_df' DataFrame.\n",
    "for index, alpha in enumerate(alphas):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_alpha_experiment()' with as \n",
    "    # argument the 'alpha'.\n",
    "    R2_scores_current_alpha = lasso_regression_with_alpha_experiment(alpha)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_alpha' to the corresponding column of the general DataFrame.\n",
    "    R2_0_to_10_df[alphas_0_to_10[index]] = R2_scores_current_alpha.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_0_to_10_df' DataFrame:\")\n",
    "R2_0_to_10_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de8d3e9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_alpha_0_to_10_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_0_to_10_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8348ed5f",
   "metadata": {},
   "source": [
    "Using the 'R2_0_to_10_df' DataFrame to which all R<sup>2</sup> scores have been added for each of the alphas, we can now create the box plots (one for each alpha) by calling the function 'boxplot()' from the 'Seaborn' library. We can also save this plot to the directory 'data_directory_results_distance' by calling the function 'savefig()'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa00a8a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(20, 12))\n",
    "\n",
    "# Creating a boxplot for every column (sample) in the 'R2_0_to_10_df' DataFrame, plotting them on the same axis, without \n",
    "# showing the outliers.\n",
    "ax = sns.boxplot(data=R2_0_to_10_df, showfliers=False)\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right', rotation_mode='anchor', fontsize=24)\n",
    "ax.tick_params(axis='y', labelsize=24)\n",
    "\n",
    "# Adding the legend, a title and the labels to the plot.\n",
    "ax.set_title('The Distributions of the R2 scores of the Alphas', pad=20, fontsize=30)\n",
    "ax.set_xlabel('Alpha', labelpad=20, fontsize=30)\n",
    "ax.set_ylabel('R2 score', labelpad=20, fontsize=30)\n",
    "\n",
    "# Saving the plot by calling the function 'savefig()'.\n",
    "file_to_save = data_directory_results_distance + f\"/R2_scores_alphas_0_to_10.png\"\n",
    "plt.savefig(file_to_save, bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed361bac",
   "metadata": {},
   "source": [
    "As we can see from the box plots above, it seems that there is a clear peak around alpha being equal to 0.10 as the median R<sup>2</sup> score decreases as alpha increases. Another thing we can observe is that from a value of alpha being equal to 1 the distributions of the R<sup>2</sup> scores barely change. In Lasso regression, an alpha value greater than 1 is often avoided as it can result in models with excessive constraints. The loss function's penalty term is stronger as alpha increases, which encourages more coefficients to be set to exactly zero. As a result, several coefficients are effectively deleted, leading to a more sparse model.\n",
    "\n",
    "To investigate what the actual best alpha is, we can define a new list of alphas ranging from 0.05 to 0.15 for which each the function 'lasso_regression_with_alpha_experiment()' is called. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0640126",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The list of alphas ranging from 0.05 through 0.15 for which each the function 'lasso_regression_with_alpha_experiment()' \n",
    "# is called.\n",
    "alphas = [0.05, 0.06, 0.07, 0.08, 0.09, 0.10, 0.11, 0.12, 0.13, 0.14, 0.15]\n",
    "\n",
    "# Creating a dictionary which will later contain the lists of R2 scores for the alpha values defined in the list above.\n",
    "alphas_005_to_015 = ['0.05', '0.06', '0.07', '0.08', '0.09', '0.10', '0.11', '0.12', '0.13', '0.14', '0.15']\n",
    "R2_005_to_015 = {alpha: [] for alpha in alphas_005_to_015}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_005_to_015_df = pd.DataFrame(R2_005_to_015)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_005_to_015_df.insert(0, 'Gene', gene_expression_data_log2_transformed_chromosome1.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_005_to_015_df' DataFrame:\")\n",
    "R2_005_to_015_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "105b4368",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looping over every 'alpha' within the 'alphas' list and applying lasso regression to the M-transformed methylation data \n",
    "# and log2-transformed gene expression data for the 'alpha' and adding their resulting lists filled with R2 scores to the \n",
    "# 'R2_005_to_015_df' DataFrame.\n",
    "for index, alpha in enumerate(alphas):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_alpha_experiment()' with as \n",
    "    # argument the 'alpha'.\n",
    "    R2_scores_current_alpha = lasso_regression_with_alpha_experiment(alpha)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_alpha' to the corresponding column of the general DataFrame.\n",
    "    R2_005_to_015_df[alphas_005_to_015[index]] = R2_scores_current_alpha.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_005_to_015_df' DataFrame:\")\n",
    "R2_005_to_015_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59459717",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_alpha_005_to_015_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_005_to_015_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596fdbf8",
   "metadata": {},
   "source": [
    "Using the 'R2_005_to_015_df' DataFrame to which all R<sup>2</sup> scores have been added for each of the alphas, we can now create the box plots (one for each alpha) by calling the function 'boxplot()' from the 'Seaborn' library. We can also save this plot to the directory 'data_directory_results_distance' by calling the function 'savefig()'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec46d521",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(20, 12))\n",
    "\n",
    "# Creating a boxplot for every column (sample) in the 'R2_015_to_025_df' DataFrame, plotting them on the same axis, without \n",
    "# showing the outliers.\n",
    "ax = sns.boxplot(data=R2_005_to_015_df, showfliers=False)\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right', rotation_mode='anchor', fontsize=17)\n",
    "ax.tick_params(axis='y', labelsize=17)\n",
    "\n",
    "# Adding the legend, a title and the labels to the plot.\n",
    "ax.set_title('The Distributions of the R2 scores of the Alphas', pad=20, fontsize=30)\n",
    "ax.set_xlabel('Alpha', labelpad=20, fontsize=30)\n",
    "ax.set_ylabel('R2 score', labelpad=20, fontsize=30)\n",
    "\n",
    "# Saving the plot by calling the function 'savefig()'.\n",
    "file_to_save = data_directory_results_distance + f\"/R2_scores_alphas_005_to_015.png\"\n",
    "plt.savefig(file_to_save, bbox_inches='tight')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9787302e",
   "metadata": {},
   "source": [
    "As we can see from the box plots above, the difference between the distributions is not that large which is expected as the difference between the alphas is only 0.01. Since the distributions look fairly similar, there would not be a big difference in performance when choosing one over the other for the next steps in the machine learning algorithms application for the Distance Analysis part. We can therefore base our decision on which of the medians of the distributions is the highest for which alpha. This is done in the code below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "02daff57",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-16T23:27:46.756492Z",
     "start_time": "2023-06-16T23:27:46.359269Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'R2_015_to_025_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-14-dc1b5dc136c3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# Calculating the median of each of the columns (alphas) of the 'R2_015_to_025_df' DataFrame by calling the function\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;31m# 'median()'.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m \u001b[0mcolumn_medians\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mR2_015_to_025_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmedian\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;31m# Retrieving the name of the column (the alpha) which has the highest median present in the 'column_medians' list by calling\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'R2_015_to_025_df' is not defined"
     ]
    }
   ],
   "source": [
    "# Calculating the median of each of the columns (alphas) of the 'R2_005_to_015_df' DataFrame by calling the function \n",
    "# 'median()'.\n",
    "column_medians = R2_005_to_015_df.median(numeric_only=True)\n",
    "\n",
    "# Retrieving the name of the column (the alpha) which has the highest median present in the 'column_medians' list by calling \n",
    "# the function 'idxmax()'.\n",
    "alpha_with_highest_median = column_medians.idxmax()\n",
    "\n",
    "print(\"Alpha with the highest median: \", alpha_with_highest_median)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4d90bcd",
   "metadata": {},
   "source": [
    "As we can see from the output above, an alpha of 0.07 has the highest median indicating that this will be the alpha used throughout the remainder of this notebook for experimenting with different distances in the section 'Applying Lasso Regression'."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c65bcaf",
   "metadata": {},
   "source": [
    "### Applying Lasso Regression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae1fba96",
   "metadata": {},
   "source": [
    "Now we can apply lasso regression with an alpha of 0.07 to the M-transformed methylation data and log2-transformed gene expression data. The distances will be divided into 8 blocks of 5 distances because when they all appear within the same code block, it is quite a memory expensive computation (and we do not want to lose any progress). Separating these distances into multiple blocks which allows us to save a different DataFrame featuring the R<sup>2</sup> scores for each of the 8 blocks which can later be easily recombined into a single DataFrame for displaying purposes. This can be achieved by calling the function 'to.csv()' for each DataFrame.\n",
    "\n",
    "Before we do this we should first define the function 'lasso_regression_with_distance()' below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae91f7c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function retrieves the R2 scores for the lasso regression models (one for each gene) fitted to predict the   \n",
    "# 'gene_expression_data_log2_transformed' based on the 'methylation_data_M_transformed'.\n",
    "def lasso_regression_with_distance(distance):\n",
    "    \n",
    "    # Retrieving the genes present in the DataFrame 'gene_expression_data_log2_transformed'.\n",
    "    genes = gene_expression_data_log2_transformed.columns[1:]\n",
    "    \n",
    "    # Defining a list where all the R2 scores (one for each gene) will be stored such that we can later represent these\n",
    "    # within a box plot to compare them with the R2 scores for the other experiments. This can be achieved by calling the\n",
    "    # function 'calculate_R2_scores()' for each of the genes. Since the computations for a single gene do not influence the \n",
    "    # computations of any other gene, we can parallelize the execution of this function by calling the function 'Parallel()' \n",
    "    # from the 'joblib' library.\n",
    "    R2_scores = Parallel(n_jobs=512)(delayed(calculate_R2_scores)(gene, distance, 0.07) for gene in genes)\n",
    "    \n",
    "    # Combining all the key-value pairs into a single dictionary.\n",
    "    R2_scores_dictionary = {R2_score['gene']: R2_score['R2'] for R2_score in R2_scores}\n",
    "    \n",
    "    return R2_scores_dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee0aad59",
   "metadata": {},
   "source": [
    "In addition, we should also define the list featuring all the distances that will be experimented with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e5db8fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining the list featuring all the distances that will be experimented with.\n",
    "distances = [5000, 10000, 15000, 25000, 50000, 75000, 100000, 150000, 250000, 350000, 500000, 750000, \n",
    "             1000000, 1500000, 2000000, 2500000, 4000000, 5000000, 6000000, 7500000, 10000000, 12500000, \n",
    "             15000000, 17500000, 20000000, 25000000, 30000000, 40000000, 50000000, 65000000, 80000000, \n",
    "             100000000, 120000000, 150000000, 200000000, 250000000, 350000000, 500000000, 750000000, 1000000000]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b25b07ed",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 5,000 through 50,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "adee64dd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:35:42.745667Z",
     "start_time": "2023-06-12T20:35:42.715489Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The empty 'R2_5000_to_50000_df' DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>5,000</th>\n",
       "      <th>10,000</th>\n",
       "      <th>15,000</th>\n",
       "      <th>25,000</th>\n",
       "      <th>50,000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000001561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000001629</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000002587</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000002746</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3438</th>\n",
       "      <td>ENSG00000287893</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439</th>\n",
       "      <td>ENSG00000288156</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3440</th>\n",
       "      <td>ENSG00000288586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3441</th>\n",
       "      <td>ENSG00000288612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3442</th>\n",
       "      <td>ENSG00000288658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3443 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene  5,000  10,000  15,000  25,000  50,000\n",
       "0     ENSG00000001561    NaN     NaN     NaN     NaN     NaN\n",
       "1     ENSG00000001629    NaN     NaN     NaN     NaN     NaN\n",
       "2     ENSG00000001631    NaN     NaN     NaN     NaN     NaN\n",
       "3     ENSG00000002587    NaN     NaN     NaN     NaN     NaN\n",
       "4     ENSG00000002746    NaN     NaN     NaN     NaN     NaN\n",
       "...               ...    ...     ...     ...     ...     ...\n",
       "3438  ENSG00000287893    NaN     NaN     NaN     NaN     NaN\n",
       "3439  ENSG00000288156    NaN     NaN     NaN     NaN     NaN\n",
       "3440  ENSG00000288586    NaN     NaN     NaN     NaN     NaN\n",
       "3441  ENSG00000288612    NaN     NaN     NaN     NaN     NaN\n",
       "3442  ENSG00000288658    NaN     NaN     NaN     NaN     NaN\n",
       "\n",
       "[3443 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 5,000 through 50,000.\n",
    "numbers_5000_to_50000 = ['5,000', '10,000', '15,000', '25,000', '50,000']\n",
    "R2_5000_to_50000 = {number: [] for number in numbers_5000_to_50000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_5000_to_50000_df = pd.DataFrame(R2_5000_to_50000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_5000_to_50000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_5000_to_50000_df' DataFrame:\")\n",
    "R2_5000_to_50000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcb1b1a9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.768077Z",
     "start_time": "2023-06-12T19:59:28.783809Z"
    }
   },
   "outputs": [],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_5000_to_50000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[0:5]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_5000_to_50000_df[numbers_5000_to_50000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_5000_to_50000_df' DataFrame:\")\n",
    "R2_5000_to_50000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8197e3d1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.771628Z",
     "start_time": "2023-06-12T19:55:40.737Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_5000_to_50000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_5000_to_50000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6503ed2",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 75,000 through 350,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eff74ffe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.772831Z",
     "start_time": "2023-06-12T19:55:40.739Z"
    }
   },
   "outputs": [],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 75,000 through 350,000.\n",
    "numbers_75000_to_350000 = ['75,000', '100,000', '150,000', '250,000', '350,000']\n",
    "R2_75000_to_350000 = {number: [] for number in numbers_75000_to_350000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_75000_to_350000_df = pd.DataFrame(R2_75000_to_350000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_75000_to_350000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_75000_to_350000_df' DataFrame:\")\n",
    "R2_75000_to_350000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e15a242",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_75000_to_350000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[5:10]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_75000_to_350000_df[numbers_75000_to_350000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_75000_to_350000_df' DataFrame:\")\n",
    "R2_75000_to_350000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c2aab2c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.774842Z",
     "start_time": "2023-06-12T19:55:40.742Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_75000_to_350000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_75000_to_350000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24853876",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 500,000 through 2,000,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d4eb0f2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.775879Z",
     "start_time": "2023-06-12T19:55:40.744Z"
    }
   },
   "outputs": [],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 500,000 through 2,000,000.\n",
    "numbers_500000_to_2000000 = ['500,000', '750,000', '1,000,000', '1,500,000', '2,000,000']\n",
    "R2_500000_to_2000000 = {number: [] for number in numbers_500000_to_2000000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_500000_to_2000000_df = pd.DataFrame(R2_500000_to_2000000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_500000_to_2000000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_500000_to_2000000_df' DataFrame:\")\n",
    "R2_500000_to_2000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7288f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_500000_to_2000000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[10:15]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_500000_to_2000000_df[numbers_500000_to_2000000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The '500000_to_2000000' DataFrame:\")\n",
    "R2_500000_to_2000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccef9781",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.776880Z",
     "start_time": "2023-06-12T19:55:40.747Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_500000_to_2000000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_500000_to_2000000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b0f879f",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 2,500,000 through 7,500,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae6304f3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.777883Z",
     "start_time": "2023-06-12T19:55:40.750Z"
    }
   },
   "outputs": [],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 2,500,000 through 7,500,000.\n",
    "numbers_2500000_to_7500000 = ['2,500,000', '4,000,000', '5,000,000', '6,000,000', '7,500,000']\n",
    "R2_2500000_to_7500000 = {number: [] for number in numbers_2500000_to_7500000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_2500000_to_7500000_df = pd.DataFrame(R2_2500000_to_7500000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_2500000_to_7500000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_2500000_to_7500000_df' DataFrame:\")\n",
    "R2_2500000_to_7500000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21da0218",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_2500000_to_7500000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[15:20]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_2500000_to_7500000_df[numbers_2500000_to_7500000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_2500000_to_7500000_df' DataFrame:\")\n",
    "R2_2500000_to_7500000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f0b4ac6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.778842Z",
     "start_time": "2023-06-12T19:55:40.753Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_2500000_to_7500000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_2500000_to_7500000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70ca799",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 10,000,000 through 20,000,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fedb5436",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.779881Z",
     "start_time": "2023-06-12T19:55:40.755Z"
    }
   },
   "outputs": [],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 10,000,000 through 20,000,000.\n",
    "numbers_10000000_to_20000000 = ['10,000,000', '12,500,000', '15,000,000', '17,500,000', '20,000,000']\n",
    "R2_10000000_to_20000000 = {number: [] for number in numbers_10000000_to_20000000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_10000000_to_20000000_df = pd.DataFrame(R2_10000000_to_20000000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_10000000_to_20000000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_10000000_to_20000000_df' DataFrame:\")\n",
    "R2_10000000_to_20000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb676b4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_10000000_to_20000000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[20:25]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_10000000_to_20000000_df[numbers_10000000_to_20000000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_10000000_to_20000000_df' DataFrame:\")\n",
    "R2_10000000_to_20000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5572cd5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.781301Z",
     "start_time": "2023-06-12T19:55:40.758Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_10000000_to_20000000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_10000000_to_20000000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8892e039",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 25,000,000 through 65,000,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c41de52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.782531Z",
     "start_time": "2023-06-12T19:55:40.760Z"
    }
   },
   "outputs": [],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 25,000,000 through 65,000,000.\n",
    "numbers_25000000_to_65000000 = ['25,000,000', '30,000,000', '40,000,000', '50,000,000', '65,000,000']\n",
    "R2_25000000_to_65000000 = {number: [] for number in numbers_25000000_to_65000000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_25000000_to_65000000_df = pd.DataFrame(R2_25000000_to_65000000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_25000000_to_65000000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_25000000_to_65000000_df' DataFrame:\")\n",
    "R2_25000000_to_65000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f557bdfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_25000000_to_65000000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[25:30]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_25000000_to_65000000_df[numbers_25000000_to_65000000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_25000000_to_65000000_df' DataFrame:\")\n",
    "R2_25000000_to_65000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c52749eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.783041Z",
     "start_time": "2023-06-12T19:55:40.762Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_25000000_to_65000000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_25000000_to_65000000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cd1efc3",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 80,000,000 through 200,000,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84c49d74",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.784067Z",
     "start_time": "2023-06-12T19:55:40.764Z"
    }
   },
   "outputs": [],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 80,000,000 through 200,000,000.\n",
    "numbers_80000000_to_200000000 = ['80,000,000', '100,000,000', '120,000,000', '150,000,000', '200,000,000']\n",
    "R2_80000000_to_200000000 = {number: [] for number in numbers_80000000_to_200000000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_80000000_to_200000000_df = pd.DataFrame(R2_80000000_to_200000000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_80000000_to_200000000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_80000000_to_200000000_df' DataFrame:\")\n",
    "R2_80000000_to_200000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "012dd6b7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-14T23:04:00.667562Z",
     "start_time": "2023-06-14T23:04:00.372183Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'distances' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-51e15cbd654b>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;31m# regression to the M-transformed methylation data and log2 transformed gene expression data for the 'distance' and adding\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;31m# their resulting lists filled with R2 scores to the 'R2_80000000_to_200000000_df' DataFrame.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[1;32mfor\u001b[0m \u001b[0mindex\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdistance\u001b[0m \u001b[1;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdistances\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m30\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;36m35\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m     \u001b[0mstart\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'distances' is not defined"
     ]
    }
   ],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_80000000_to_200000000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[30:35]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_80000000_to_200000000_df[numbers_80000000_to_200000000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_80000000_to_200000000_df' DataFrame:\")\n",
    "R2_80000000_to_200000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45c48e85",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.787511Z",
     "start_time": "2023-06-12T19:55:40.767Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_80000000_to_200000000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_80000000_to_200000000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a321fb89",
   "metadata": {},
   "source": [
    "#### Lasso Regression for Distances 250,000,000 through 1,000,000,000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1b4c812",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.788522Z",
     "start_time": "2023-06-12T19:55:40.769Z"
    }
   },
   "outputs": [],
   "source": [
    "# Creating a dictionary which will later contain the lists of R2 scores for distance 250,000,000 through 1,000,000,000.\n",
    "numbers_250000000_to_1000000000 = ['250,000,000', '350,000,000', '500,000,000', '750,000,000', '1,000,000,000']\n",
    "R2_250000000_to_1000000000 = {number: [] for number in numbers_250000000_to_1000000000}\n",
    "\n",
    "# Creating the DataFrame by calling the constructor 'DataFrame()' which takes as input the dictionary to be converted into a\n",
    "# DataFrame.\n",
    "R2_250000000_to_1000000000_df = pd.DataFrame(R2_250000000_to_1000000000)\n",
    "\n",
    "# Adding the names of the all the genes for which the R2 score is computed to the DataFrame storing all of the R2 scores.\n",
    "R2_250000000_to_1000000000_df.insert(0, 'Gene', gene_expression_data_log2_transformed.columns[1:])\n",
    "\n",
    "print(\"The empty 'R2_250000000_to_1000000000_df' DataFrame:\")\n",
    "R2_250000000_to_1000000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3438a4b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looping over every 'distance' within the 'distances' list that is present in the current block and applying lasso \n",
    "# regression to the M-transformed methylation data and log2-transformed gene expression data for the 'distance' and adding \n",
    "# their resulting lists filled with R2 scores to the 'R2_250000000_to_1000000000_df' DataFrame.\n",
    "for index, distance in enumerate(distances[35:]):\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    # Retrieving the R2 scores for the lasso regression models (one for each gene) fitted to predict the gene expression \n",
    "    # values based on the methylation data by calling the function 'lasso_regression_with_distance()' with as argument the \n",
    "    # 'distance' which will be used to filter the CpG sites used to predict the expression value for a gene.\n",
    "    R2_scores_current_distance = lasso_regression_with_distance(distance)\n",
    "    \n",
    "    # Adding the 'R2_scores_current_distance' to the corresponding column of the general DataFrame.\n",
    "    R2_250000000_to_1000000000_df[numbers_250000000_to_1000000000[index]] = R2_scores_current_distance.values()\n",
    "    \n",
    "    end = time.time()\n",
    "    print(f\"{end-start} seconds\")\n",
    "    \n",
    "print(\"The 'R2_250000000_to_1000000000_df' DataFrame:\")\n",
    "R2_250000000_to_1000000000_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d4169cc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-12T20:31:29.790512Z",
     "start_time": "2023-06-12T19:55:40.772Z"
    }
   },
   "outputs": [],
   "source": [
    "# Defining where to save the resulting file and its name.\n",
    "file_to_save = data_directory_results_distance + \"/R2_250000000_to_1000000000_df.csv\"\n",
    "\n",
    "# Saving the file.\n",
    "if not os.path.exists(file_to_save):\n",
    "    R2_250000000_to_1000000000_df.to_csv(file_to_save, index=False)\n",
    "    print(\"The file with the path \" + file_to_save + \" has been created.\")\n",
    "else:\n",
    "    print(\"The file with the path \" + file_to_save + \" already exists.\")"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
